components/properties/src/provider/bidi_data.rs

```diff
@@ -0,0 +1,236 @@
+#[repr(packed)]
+pub struct MirroredPairedBracketDataULE([u8; 3]);
+
+// Safety (based on the safety checklist on the ULE trait):
+//  1. MirroredPairedBracketDataULE does not include any uninitialized or padding bytes
+//     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant)
+//  2. MirroredPairedBracketDataULE is aligned to 1 byte.
+//     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant)
+//  3. The impl of validate_byte_slice() returns an error if any byte is not valid.
+//  4. The impl of validate_byte_slice() returns an error if there are extra bytes.
+//  5. The other ULE methods use the default impl.
+//  6. MirroredPairedBracketDataULE byte equality is semantic equality because all bits
+//     are used, so no unused bits requires no extra work to zero out unused bits
+unsafe impl ULE for MirroredPairedBracketDataULE {
+    #[inline]
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), ZeroVecError> {
+        if bytes.len() % 3 != 0 {
+            return Err(ZeroVecError::length::<Self>(bytes.len()));
+        }
+        // Validate the bytes
+        #[allow(clippy::indexing_slicing)] // Won't panic because the chunks are always 3 bytes long
+        for byte_triple in bytes.chunks_exact(3) {
+            // Bidi_Mirroring_Glyph validation
+            #[allow(clippy::unwrap_used)] // chunks_exact returns slices of length 3
+            let [byte0, byte1, byte2] = *<&[u8; 3]>::try_from(byte_triple).unwrap();
+            let mut mirroring_glyph_code_point: u32 = (byte2 & 0x1F) as u32;
+            mirroring_glyph_code_point = (mirroring_glyph_code_point << 8) | (byte1 as u32);
+            mirroring_glyph_code_point = (mirroring_glyph_code_point << 8) | (byte0 as u32);
+            let _mirroring_glyph =
+                char::from_u32(mirroring_glyph_code_point).ok_or(ZeroVecError::parse::<Self>())?;
+
+            // skip validating the Bidi_Mirrored boolean since it is always valid
+
+            // assert that Bidi_Paired_Bracket_Type cannot have a 4th value because it only
+            // has 3 values: Open, Close, None
+            if (byte2 & 0xC0) == 0xC0 {
+                return Err(ZeroVecError::parse::<Self>());
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl AsULE for MirroredPairedBracketData {
+    type ULE = MirroredPairedBracketDataULE;
+
+    #[inline]
+    fn to_unaligned(self) -> Self::ULE {
+        let mirr_glyph_char_ule_ref_slice = &[self.mirroring_glyph.to_unaligned()];
+        let byte_slice = CharULE::as_byte_slice(mirr_glyph_char_ule_ref_slice);
+        let mut byte2 = byte_slice.get(2).copied().unwrap_or_default();
+        byte2 |= (self.mirrored as u8) << 5;
+        byte2 |= self.paired_bracket_type.0 << 6;
+
+        MirroredPairedBracketDataULE([
+            byte_slice.first().copied().unwrap_or_default(),
+            byte_slice.get(1).copied().unwrap_or_default(),
+            byte2,
+        ])
+    }
```

Comment:
Suggestion: re-write this method as

```suggestion
    fn to_unaligned(self) -> Self::ULE {
        let ch = u32::from(self.mirroring_glyph);
        ch |= u32::from(self.mirrored) << 21;
        ch |= self.paired_bracket_type.0 << 22;
        let [byte0, byte1, byte2, _] = ch.to_le_bytes();
        MirroredPairedBracketDataULE([byte0, byte1, byte2])
    }
```